doc/source/debugging.rst

```diff
@@ -1,54 +1,183 @@
+
+Formatter
+---------
+
+.. autoclass:: bluesky.log.LogFormatter
+
+Global Handler
+---------------
+
+Following Python's recommendation, bluesky does not install any handlers at
+import time, but it provides a function to set up a basic useful configuration
+in one line, similar to Python's :py:func:`logging.basicConfig` but with some
+additional options---and scoped to the ``'bluesky'`` logger with bluesky's
+:class:`bluesky.log.LogFormatter`. It streamlines common use cases without
+interfering with more sophisticated use cases.
+
+.. autofunction:: bluesky.log.config_bluesky_logging
+.. autofunction:: bluesky.log.get_handler
+
+Advanced Example
+================
+
+The flow of log event information in loggers and handlers is illustrated in the
+following diagram:
+
+.. image:: https://docs.python.org/3/_images/logging_flow.png
+
+For further reference, see the Python 3 logging howto:
+https://docs.python.org/3/howto/logging.html#logging-flow
+
+As an illustrative example, we will set up two handlers using the Python
+logging framework directly, ignoring bluesky's convenience function.
+
+Suppose we set up a handler aimed at a file:
+
+.. code-block:: python
+
+    import logging
+    file_handler = logging.FileHandler('bluesky.log')
+
+And another aimed at `Logstash <https://www.elastic.co/products/logstash>`_:
+
+.. code-block:: python
+
+    import logstash  # requires python-logstash package
+    logstash_handler = logstash.TCPLogstashHandler(<host>, <port>, version=1)
+
+We can attach the handlers to the bluesky logger, to which all log records
+created by bluesky propagate:
+
+.. code-block:: python
+
+    logger = logging.getLogger('bluesky')
+    logger.addHandler(logstash_handler)
+    logger.addHandler(file_filter)
+
+We can set the verbosity of each handler. Suppose want maximum verbosity in the
+file but only medium verbosity in logstash.
+
+.. code-block:: python
+
+    logstash_handler.setLevel('INFO')
+    file_handler.setLevel('DEBUG')
+
+Finally, ensure that "effective level" of ``logger`` is at least as verbose as
+the most verbose handler---in this case, ``'DEBUG'``. By default, at import,
+its level is not set
+
+.. ipython:: python
+   :verbatim:
+
+    logging.getLevelName(logger.level)
+    'NOTSET'
+
+and so it inherits the level of Python's default
+"handler of last resort," :py:obj:`logging.lastResort`, which is ``'WARNING'``.
+
+.. ipython:: python
+   :verbatim:
+
+    logging.getLevelName(logger.getEffectiveLevel())
+    'WARNING'
+
+In this case we should set it to ``'DEBUG'``, to match the most verbose level
+of the handler we have added.
+
+.. code-block:: python
+
+   logger.setLevel('DEBUG')
+
+This makes DEBUG-level records *available* to all handlers. Our logstash
+handler, set to ``'INFO'``, will filter out DEBUG-level records.
+
+To globally disable the generation of any log records at or below a certain
+verbosity, which may be helpful for optimising performance, Python provides
```

Comment:
Are you practicing your Canadian in case Trump is re-elected? :)

Reply:
I honestly think that reading @stuartcampbell's writing has finally got to me. Well done, :gb:.